src/System.IO.Pipelines.Networking.Tls/SecurityContext.cs

```diff
@@ -0,0 +1,196 @@
+        /// <param name="hostName">The name of the host that will be sent to the other parties, for a server this should be the name on the certificate. For clients this can be left blank or the name on a client cert</param>
+        /// <param name="isServer">Used to denote if you are going to be negotiating incoming or outgoing Tls connections</param>
+        /// <param name="serverCert">This is the in memory representation of the certificate used for the PKI exchange and authentication</param>
+        /// <param name="alpnSupportedProtocols">This is the protocols that are supported and that will be negotiated with on the otherside, if a protocol can't be negotiated then the handshake will fail</param>
+        public unsafe SecurityContext(PipelineFactory factory, string hostName, bool isServer,
+            X509Certificate serverCert, ApplicationProtocols.ProtocolIds alpnSupportedProtocols)
+        {
+            if (hostName == null)
+            {
+                throw new ArgumentNullException(nameof(hostName));
+            }
+            _hostName = hostName;
+            _pipelineFactory = factory;
+            _serverCertificate = serverCert;
+            _isServer = isServer;
+            CreateAuthentication();
+            if (alpnSupportedProtocols > 0)
+            {
+                //We need to get a buffer for the ALPN negotiation and pin it for sending to the lower API
+                _alpnSupportedProtocols = ApplicationProtocols.GetBufferForProtocolId(alpnSupportedProtocols, true);
+                _alpnHandle = GCHandle.Alloc(_alpnSupportedProtocols, GCHandleType.Pinned);
+                _alpnBuffer = new SecurityBuffer((void*) _alpnHandle.AddrOfPinnedObject(),
+                    _alpnSupportedProtocols.Length, SecurityBufferType.ApplicationProtocols);
+            }
+        }
+
+        internal SSPIHandle CredentialsHandle => _credsHandle;
+        internal bool AplnRequired => _alpnSupportedProtocols != null;
+        internal SecurityBuffer AplnBuffer => _alpnBuffer;
+        internal string HostName => _hostName;
+        public bool IsServer => _isServer;
+
+        private unsafe void CreateAuthentication()
+        {
+            int numberOfPackages;
+            SecPkgInfo* secPointer = null;
+            try
+            {
+                //Load the available security packages and look for the Unified pack from MS that supplies TLS support
+                if (Interop.EnumerateSecurityPackagesW(out numberOfPackages, out secPointer) != 0)
+                {
+                    throw new InvalidOperationException("Unable to enumerate security packages");
+                }
+                for (var i = 0; i < numberOfPackages; i++)
+                {
+                    var package = secPointer[i];
+                    var name = Marshal.PtrToStringUni(package.Name);
```

Comment:
even if you don't know the size ahead of time? does that work if so great.. I wasn't sure.